grpc-gateway/subscription/subscription.go

```diff
@@ -0,0 +1,840 @@
+	var e events.DeviceMetadataUpdated
+	if err := eu.Unmarshal(&e); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal event %v: %w", eu, err)
+	}
+	return &pb.Event{
+		Type: &pb.Event_DeviceMetadataUpdated{
+			DeviceMetadataUpdated: &e,
+		},
+	}, nil
+}
+
+var eventToHandler = map[string]resourceEventHandler{
+	(&events.ResourceCreatePending{}).EventType():       handleResourceCreatePending,
+	(&events.ResourceCreated{}).EventType():             handleResourceCreated,
+	(&events.ResourceRetrievePending{}).EventType():     handleResourceRetrievePending,
+	(&events.ResourceRetrieved{}).EventType():           handleResourceRetrieved,
+	(&events.ResourceUpdatePending{}).EventType():       handleResourceUpdatePending,
+	(&events.ResourceUpdated{}).EventType():             handleResourceUpdated,
+	(&events.ResourceDeletePending{}).EventType():       handleResourceDeletePending,
+	(&events.ResourceDeleted{}).EventType():             handleResourceDeleted,
+	(&events.DeviceMetadataUpdatePending{}).EventType(): handleDeviceMetadataUpdatePending,
+	(&events.DeviceMetadataUpdated{}).EventType():       handleDeviceMetadataUpdated,
+	(&events.ResourceChanged{}).EventType():             handleResourceChanged,
+	(&events.ResourceLinksPublished{}).EventType():      handleResourcesPublished,
+	(&events.ResourceLinksUnpublished{}).EventType():    handleResourcesUnpublished,
+}
+
+func (s *Sub) handleEvent(eu eventbus.EventUnmarshaler) {
+	if !s.isFiltered(eu) {
+		return
+	}
+	handler, ok := eventToHandler[eu.EventType()]
+	if !ok {
+		log.Errorf("unhandled event type %v", eu.EventType())
+		return
+	}
+
+	ev, err := handler(eu)
+	if err != nil {
+		log.Errorf("cannot get event: %w", err)
+		return
+	}
+	ev.CorrelationId = s.correlationID
+	ev.SubscriptionId = s.id
+	err = s.send(ev)
+	if err != nil {
+		log.Errorf("cannot send event %v: %w", ev, err)
+	}
+}
+
+func (s *Sub) isFiltered(ev event) bool {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	if s.isDuplicatedInitEventLocked(ev) {
+		return false
+	}
+	if s.closed {
+		return false
+	}
+	return true
+}
+
+func (s *Sub) Handle(ctx context.Context, iter eventbus.Iter) (err error) {
+	for {
+		eu, ok := iter.Next(ctx)
+		if !ok {
+			break
+		}
+		if !isFilteredResourceIDs(s.filteredResourceIDs, eu.AggregateID()) {
+			continue
+		}
+		if !isFilteredEventype(s.filter, eu.EventType()) {
+			continue
+		}
+
+		select {
+		case <-s.done:
+			return nil
+		case s.eventsCache <- eu:
+		}
+	}
+
+	return iter.Err()
+}
+
+func (s *Sub) SetContext(ctx context.Context) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	s.ctx = ctx
+}
+
+func (s *Sub) Id() string {
+	return s.id
+}
+
+func (s *Sub) initOwnerSubscription(ownerCache *client.OwnerCache) ([]string, error) {
+	owner, err := grpc.OwnerFromTokenMD(s.ctx, ownerCache.OwnerClaim())
+	if err != nil {
+		return nil, grpc.ForwardFromError(codes.InvalidArgument, err)
+	}
+	s.mutex.Lock()
```

Comment:
What's being locked by `s.mutex` here? It doesn't seem necessary to lock the `ownerCache.Subscribe`, so can't this be moved after the `Subscribe` call?